### ISSUE TITLE
Backend deploy / scaling / CI + CD

### ISSUE THREAD

**Issue by batpad** [IFRCGo/go-api#1531]
Right now the backend deploy is a manual process, which involves:

 - Pushing to Github triggers a CI process to build a docker image and push it to Dockerhub.
 - We have a single VM running on Azure to run GO API - the deployment process involves ssh-ing into this VM, pulling in the new docker container and restarting the Docker process.
 - The various cronjobs we have are setup in the `crontab` of the same container and run on the same container that runs the API, this is non-ideal.
 - We have no way to scale the API container or run multiple replicas of the container in cases of high load, etc.

We should move to a modern deployment and infrastructure management workflow, where we specify the resources we need in a declarative fashion and deploy those resources as part of Github Actions in response to pushes to the `develop` and `master` branches. We should also use a container orchestration system that will allow us to deploy multiple containers and allow us to scale up and scale down resource usage as per needs.

We should:

 - declare all our infrastructure as Terraform templates: https://www.terraform.io/
 - Configure Github Actions to deploy and update infrastructure on code pushes, which will handle building the docker images needed and deploy them to the infrastructure.

We need to move away from deploying onto a single Virtual Machine, and have more flexibility in terms of scaling up and down infrastructure, and have different processes run in isolated containers. We likely have a few options to do this:

 - Azure has a few different options to run containers in a "managed way": https://azure.microsoft.com/en-us/products/category/containers/ . We can likely use a combination of Azure managed services for running services (the API) and jobs (cronjobs / management commands) and spin up and down containers on demand.

 - We do not have much familiarity with the proprietary Azure services for container management, but what we do have experience with is using Azure's managed Kubernetes service, which would allow us to do all the things we need: i.e. scale on demand, deploy in a declarative fashion, etc.

So, I think the proposal here would be:

 - Write terraform templates to create and manage an Azure Kubernetes cluster.
 - Write Kubernetes YAML templates (perhaps using `Helm` for templating: https://helm.sh) to deploy the different services and cronjobs that are needed, and specify resource requirements for each, and setup Autoscaling on the cluster so that virtual machines are created and destroyed on demand based on usage.

In the future, this will also allow us to fairly easily configure auto-scaling rules to scale up and down based on metrics we can configure like CPU / Memory usage, or Number of Requests, etc.

Both @geohacker and I have done similar things a fair bit in the past couple of years and I think if we can figure out prioritization, feel pretty confident we could get this done in a reasonable time-frame.

cc @thenav56 @tnagorra @szabozoltan69 @geohacker @tovari

**Comment by batpad** [comment]
The structure we will follow for this:

 - Add a `deploy` folder in the `go-api` repository which will contain all the code to manage the infrastructure and deploy. We considered creating a separate repository just to manage the infrastructure and deploy, but felt it's likely easier to manage and more practical to have it sit with the API / backend code.
 - `deploy/terraform/` will create terraform resource definitions to create a production and staging cluster for the IFRC GO backend.
 - `deploy/helm/` will create a Helm chart to create all the services and containers required for the GO backend.

We will then setup a CI process to build required images and run the terraform actions to update deploys using Github Actions. i.e. create workflows in `.github/workflows/` - we will store secrets that we need to manage deploys using Github Secrets.

**Comment by batpad** [comment]
So, we finally have the API running deployed via Github Actions and running on a Kubernetes cluster on Azure at https://ifrcgo-test.ds.io/ . Still a lot of things to do, enumerating some of them:

 - [ ] Cleanup hard-coded params / things added to help debugging while deploying current version:
     - Remove hard-coded API_FQDN, get it from the correct place in GH Actions
     - Look at all variables and place them appropriately
 - [ ] Start adding Kubernetes Job definitions for each of the cron jobs that need to run.
 - [ ] Add additional services: Redis, Celery (?) - cc @thenav56 might need your help here.
 - [ ] Populate secrets with correct variables to test jobs, etc.
 - [ ] Coordinate with @szabozoltan69 to start testing this on the IFRC infrastructure - we can probably start doing this in parallel as the other tasks
 - [ ] Add secrets for staging, test things work
 - [ ] Add production secrets and plan production migration (will make a separate ticket for this)
 - [ ] Document all the things.

**Comment by thenav56** [comment]
@batpad
We will need a Redis and containers running celery (other than command everything should be the same as the base go-API container).

2 separate containers will be ideal for light and heavy tasks. We can use heavy tasks container for cronjobs.
@szabozoltan69 Correct me if I am wrong here.
```yaml
  celery-default:
    <<: *base_server_setup
    command: |
      celery -A main worker \
        -Q default \
        -l info

  celery-heavy:
    <<: *base_server_setup
    command: |
      celery -A main worker \
        -Q heavy \
        --max-tasks-per-child 5 \
        -l info
```
